addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/message/BinRpcMessage.java

```diff
@@ -82,7 +81,9 @@ public BinRpcMessage(InputStream is, boolean methodHeader, String encoding) thro
             throw new EOFException("Only " + length + " bytes received reading message length");
         }
         int datasize = (new BigInteger(ArrayUtils.subarray(sig, 4, 8))).intValue();
-        byte[] message = ArrayUtils.addAll(sig, IOUtils.toByteArray(is, datasize));
+        byte payload[] = new byte[datasize];
```

Comment:
Imho, it feels less safe because the Util does way more (error handling)

Reply:
Just reading the documentation again, this is just incorrect:

> An attempt is made to read as many as len bytes, but a smaller number may be read. The number of bytes actually read is returned as an integer.

Reply:
There is good reason the implementation is like this:

```java
809    public static byte[] toByteArray(final InputStream input, final int size) throws IOException {
810
811        if (size < 0) {
812            throw new IllegalArgumentException("Size must be equal or greater than zero: " + size);
813        }
814
815        if (size == 0) {
816            return new byte[0];
817        }
818
819        final byte[] data = new byte[size];
820        int offset = 0;
821        int read;
822
823        while (offset < size && (read = input.read(data, offset, size - offset)) != EOF) {
824            offset += read;
825        }
826
827        if (offset != size) {
828            throw new IOException("Unexpected read size. current: " + offset + ", expected: " + size);
829        }
830
831        return data;
832    }
```

Reply:
I see what you mean. Nevertheless, the problem also applies to the previous read() calls in the method, which do not use the util method. So, do you think it would be sufficient to add the same safeguard `if (length != datasize) throw ...` to this method?
Or do you think it is worth keeping the dependency around for the added safety? In that case, I would at least also use `toByteArray()` for the other two reads on the stream.

Reply:
I think that for a length of 4 its way more likely to be read in one buffer than for a complete `datasize` (without knowledge on my side what a likely size is).

I would suggest to use the full blown correct implementation so that means running inside a loop and collecting the buffer potentially in multiple steps?

Reply:
Although I am not familiar with the BinRpc protocol myself, it is probably safe to assume that the message can be significantly bigger than 4 byte. So your suggestion is probably a good way of handling it. Should I create a new PR with those changes, since this one is already merged?

Reply:
That would be great!

Reply:
Here is the PR: #3202